### PR TITLE
Cvsl 172 populate user details in session

### PR DESCRIPTION
**Pull request by timharrison-moj**
This PR:
* Requires that the latest delius-wiremock server is deployed alongside it.
* Uses the `populateCurrentUser` middeware function to get and store user information in their session
* Merges this user information from Redis into `res.locals.user` for each request (instead of looking it up again)
* For prison users it stores - name, caseload, displayName, activeCaseload (+ email address from hmpps-auth)
* For probation users it stores - name, staffId, staffCode, displayName, telephoneNumber and emailAddress
* For other users, it retrieves and stores just name, displayName and emailAddress from hmpps-auth
* Alters the two current usages of this data in approval cases and probation caseload.
* Updates all unit and integration tests in line with these changes.

This is quite a big PR and fiddly so didn't want to make it bigger ... hence a couple of follow-ons:

* There were no unit tests around populateCurrentUser before and none now - on my list to put some in
* The user service has no unit tests around `getProbationUser`  - the structure of the CommunityApiClient needs to alter to make it more suitable for mocking in unit tests - on my list.

### CODE REVIEW

File: integration_tests/pages/index.ts [hunk 1]
```diff
@@ -12,7 +12,6 @@ export default class IndexPage extends Page {
   }
 
   clickCreateALicence = (): CaseloadPage => {
-    cy.task('stubGetStaffDetails')
     cy.task('stubGetManagedOffenders')
```
Comment (timharrison-moj) [inline integration_tests/pages/index.ts:15]: This is no longer retrieved in the caseload page - already present in the session and passed into the service now.

File: server/@types/CvlUserDetails.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  // probation area code
+  probationArea?: string
+
+  // Probation area description
+  probationAreaDescription?: string
+
+  // Email address - from Nomis, Delius or Auth depending on source
+  emailAddress?: string
+}
+
```
Comment (timharrison-moj) [inline server/@types/CvlUserDetails.ts:38]: The object stored in the session, representing a superset of Nomis, Delius and Auth user details.

File: server/middleware/authorisationMiddleware.ts [hunk 3]
```diff
@@ -9,21 +9,25 @@ const isAuthorisedRole = (role: string): boolean =>
...
 export default function authorisationMiddleware(req: Request, res: Response, next: NextFunction): void {
-  if (res.locals && res.locals.user && res.locals.user.token) {
-    const { authorities: roles = [] } = jwtDecode(res.locals.user.token) as { authorities?: string[] }
+  if (res.locals?.user?.token) {
+    let roles = res.locals.user?.userRoles
+    if (!roles) {
+      const { authorities: tokenRoles = [] } = jwtDecode(res.locals.user.token) as { authorities?: string[] }
+      roles = tokenRoles
+    }
+
```
Comment (timharrison-moj) [inline server/middleware/authorisationMiddleware.ts:18]: Belt and braces here - the useRoles should always be populated.

File: server/utils/utils.ts [hunk 4]
```diff
@@ -100,6 +100,11 @@ const jsonDtTo12HourTime = (dt: string): string => {
   return momentTime.isValid() ? momentTime.format('hh:mm a') : null
 }
 
+const removeDuplicates = (list: string[]): string[] => {
+  const deDuplicatedList = [...new Set(list.map(s => s))].map(s => s)
```
Comment (smcveigh941) [inline server/utils/utils.ts:104]: What does `.map()` do here? Isn't this enough?

```
return [...new Set(list)]
```

File: server/services/index.ts [hunk 5]
```diff
@@ -5,9 +5,11 @@ import CommunityService from './communityService'
 import HmppsAuthClient from '../data/hmppsAuthClient'
 import TokenStore from '../data/tokenStore'
 import CaseloadService from './caseloadService'
+import PrisonApiClient from '../data/prisonApiClient'
 
 const hmppsAuthClient = new HmppsAuthClient(new TokenStore())
-const userService = new UserService(hmppsAuthClient)
+const prisonApiClient = new PrisonApiClient('no admin token required for user service')
```
Comment (smcveigh941) [inline server/services/index.ts:11]: Putting this string as a parameter will actually result in the requests to prison API having this header?

`Authorization: Bearer no admin token required for user service`

File: server/data/hmppsAuthClient.ts [hunk 6]
```diff
@@ -32,46 +46,36 @@ function getSystemClientTokenFromHmppsAuth(username?: string): Promise<superagen
...
-
-export interface UserRole {
-  roleCode: string
-}
-
 export default class HmppsAuthClient {
   constructor(private readonly tokenStore: TokenStore) {}
 
-  private restClient(token: string): RestClient {
+  public restClient(token: string): RestClient {
```
[the remark this answers is not in the document]
Comment (timharrison-moj) [reply server/data/hmppsAuthClient.ts:52]: Yeah I've been confused by the different implementation of these, and it really needs sorting out so they are all consistent, can be mocked easily, and can operate with either an admin client token or the user's own token when necessary. I've been mulling this over, but didn't want to extend this work any further. 3 days and counting...

File: server/utils/utils.ts (hunk 4 above)
Comment (timharrison-moj) [reply server/utils/utils.ts:104]: I think you're right. I've got some tests around this so can try a few alternatives in the morning.

**Pull request merged**